File: cli/cmd/versions.go
package cmd

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/alperen-albayrak/kutup/cli/internal/api"
	"github.com/alperen-albayrak/kutup/cli/internal/crypto"
	"github.com/spf13/cobra"
)

var versionsCmd = &cobra.Command{
	Use:   "versions",
	Short: "List, download, restore, and label snapshot versions of a file",
}

var versionsListCmd = &cobra.Command{
	Use:   "list <file-id>",
	Short: "List snapshot versions of a file (newest first)",
	Args:  cobra.ExactArgs(1),
	RunE:  runVersionsList,
}

var versionsDownloadCmd = &cobra.Command{
	Use:   "download <file-id> <version-id> [dest]",
	Short: "Download a specific snapshot version",
	Args:  cobra.RangeArgs(2, 3),
	RunE:  runVersionsDownload,
}

var versionsRestoreCmd = &cobra.Command{
	Use:   "restore <file-id> <version-id>",
	Short: "Restore a snapshot as the latest version (creates a new snapshot)",
	Args:  cobra.ExactArgs(2),
	RunE:  runVersionsRestore,
}

var (
	versionLabelKeepForever bool
)

var versionsLabelCmd = &cobra.Command{
	Use:   "label <file-id> <version-id> <label>",
	Short: "Set a label on a version (and optionally pin it via --keep-forever)",
	Args:  cobra.ExactArgs(3),
	RunE:  runVersionsLabel,
}

func init() {
	versionsLabelCmd.Flags().BoolVar(&versionLabelKeepForever, "keep-forever", false,
		"pin the version so version_cleanup never expires it")

	versionsCmd.AddCommand(versionsListCmd)
	versionsCmd.AddCommand(versionsDownloadCmd)
	versionsCmd.AddCommand(versionsRestoreCmd)
	versionsCmd.AddCommand(versionsLabelCmd)
}

func runVersionsList(_ *cobra.Command, args []string) error {
	fileID := args[0]
	client, _, cleanup, err := requireSessionFull()
	if err != nil {
		return err
	}
	defer cleanup()

	versions, err := client.ListVersions(fileID)
	if err != nil {
		return err
	}

	if jsonOut {
		out, _ := json.Marshal(versions)
		fmt.Println(string(out))
		return nil
	}
	if len(versions) == 0 {
		fmt.Println("(no snapshot versions for this file)")
		return nil
	}
	fmt.Printf("%-36s  %-20s  %12s  %s  %s\n", "ID", "CREATED", "SIZE", "PIN", "LABEL")
	for _, v := range versions {
		label := ""
		if v.Label != nil {
			label = *v.Label
		}
		pin := " "
		if v.KeepForever {
			pin = "★"
		}
		fmt.Printf("%-36s  %-20s  %12d  %s   %s\n",
			v.ID, v.CreatedAt.Local().Format(time.RFC3339), v.SizeBytes, pin, label)
	}
	return nil
}

func runVersionsDownload(_ *cobra.Command, args []string) error {
	fileID := args[0]
	versionID := args[1]
	destDir := "."
	if len(args) == 3 {
		destDir = args[2]
	}

	client, sess, cleanup, err := requireSessionFull()
	if err != nil {
		return err
	}
	defer cleanup()

	masterKey, err := sess.MasterKeyBytes()
	if err != nil {
		return err
	}

	// Find the file in collections to unwrap its file_key.
	row, fileKey, err := findFileAndKey(client, sess, masterKey, fileID)
	if err != nil {
		return err
	}
	metaBytes, err := crypto.SecretBoxOpenB64(row.EncryptedMetadata, row.MetadataNonce, fileKey)
	if err != nil {
		return fmt.Errorf("decrypt metadata: %w", err)
	}
	var meta struct {
		Name string `json:"name"`
	}
	_ = json.Unmarshal(metaBytes, &meta)

	encrypted, err := client.DownloadVersion(fileID, versionID)
	if err != nil {
		return err
	}
	plain, err := crypto.DecryptStream(encrypted, fileKey)
	if err != nil {
		return fmt.Errorf("decrypt: %w", err)
	}

	destPath := destDir
	if fi, err := os.Stat(destDir); err == nil && fi.IsDir() {
		destPath = filepath.Join(destDir, fmt.Sprintf("%s.v-%s", meta.Name, versionID[:8]))
	}
	if err := os.WriteFile(destPath, plain, 0644); err != nil {
		return err
	}
	if jsonOut {
		fmt.Printf(`{"fileId":%q,"versionId":%q,"size":%d,"dest":%q}`+"\n",
			fileID, versionID, len(plain), destPath)
	} else {
		fmt.Printf("Downloaded version %s of %s → %s\n", versionID[:8], meta.Name, destPath)
	}
	return nil
}

func runVersionsRestore(_ *cobra.Command, args []string) error {
	fileID := args[0]
	versionID := args[1]

	client, sess, cleanup, err := requireSessionFull()
	if err != nil {
		return err
	}
	defer cleanup()

	masterKey, err := sess.MasterKeyBytes()
	if err != nil {
		return err
	}

	_, fileKey, err := findFileAndKey(client, sess, masterKey, fileID)
	if err != nil {
		return err
	}

	// Mirror frontend/src/pages/FileEditorPage.tsx:325-358 performBlobRestore:
	// download chosen version → decrypt → re-encrypt → POST snapshot-blob
	// → recordSnapshot with "Restored from <date>" label.
	encrypted, err := client.DownloadVersion(fileID, versionID)
	if err != nil {
		return err
	}
	old, err := crypto.DecryptStream(encrypted, fileKey)
	if err != nil {
		return fmt.Errorf("decrypt: %w", err)
	}
	reEncrypted, err := crypto.EncryptStream(old, fileKey)
	if err != nil {
		return fmt.Errorf("re-encrypt: %w", err)
	}
	blobRes, err := client.UploadSnapshotBlob(fileID, reEncrypted)
	if err != nil {
		return err
	}
	res, err := client.RecordSnapshot(fileID, api.RecordSnapshotRequest{
		S3VersionID:   blobRes.S3VersionID,
		StoragePath:   blobRes.StoragePath,
		SeqAtSnapshot: 0,
		DocKeyID:      1,
		SizeBytes:     int64(len(reEncrypted)),
		Label:         "Restored from " + time.Now().Local().Format(time.RFC3339),
	})
	if err != nil {
		return err
	}
	if jsonOut {
		fmt.Printf(`{"fileId":%q,"newVersionId":%q,"restoredFrom":%q}`+"\n",
			fileID, res.ID, versionID)
	} else {
		fmt.Printf("Restored: file=%s new-version=%s (from %s)\n", fileID, res.ID, versionID[:8])
	}
	return nil
}

func runVersionsLabel(_ *cobra.Command, args []string) error {
	fileID := args[0]
	versionID := args[1]
	label := args[2]

	client, _, cleanup, err := requireSessionFull()
	if err != nil {
		return err
	}
	defer cleanup()

	patch := api.PatchVersionRequest{Label: &label}
	if versionLabelKeepForever {
		t := true
		patch.KeepForever = &t
	}
	row, err := client.PatchVersion(fileID, versionID, patch)
	if err != nil {
		return err
	}
	if jsonOut {
		out, _ := json.Marshal(row)
		fmt.Println(string(out))
	} else {
		fmt.Printf("Labeled version %s of file %s\n", versionID[:8], fileID)
	}
	return nil
}

// findFileAndKey locates a file row by ID across all accessible
// collections and returns the row + unwrapped file_key. Used by versions
// download / restore where the file's metadata + key are needed.
func findFileAndKey(client *api.Client, _ interface{}, masterKey []byte, fileID string) (*api.File, []byte, error) {
	cols, err := client.ListCollections()
	if err != nil {
		return nil, nil, err
	}
	for _, col := range cols {
		// Decrypt the collection key with the master key. Assumes the
		// collection isn't a federated incoming share — federation
		// versions/restore is out of scope for PR-α.
		colKeyEnc, _ := base64.StdEncoding.DecodeString(col.EncryptedKey)
		_ = colKeyEnc
		// Reuse the existing helper.
		colKey, err := crypto.SecretBoxOpenB64(col.EncryptedKey, col.EncryptedKeyNonce, masterKey)
		if err != nil {
			continue
		}
		files, err := client.ListFiles(col.ID)
		if err != nil {
			continue
		}
		for _, f := range files {
			if f.ID != fileID {
				continue
			}
			fk, err := crypto.SecretBoxOpenB64(f.EncryptedFileKey, f.FileKeyNonce, colKey)
			if err != nil {
				return nil, nil, fmt.Errorf("unwrap file key: %w", err)
			}
			return &f, fk, nil
		}
	}
	return nil, nil, fmt.Errorf("file %s not found in any accessible collection", fileID)
}

File: crates/kutup-cli/src/commands/versions.rs
//! `kutup versions` — list/download/restore/label file snapshots.
//! Mirrors `cmd/versions.go`.

use anyhow::{Context, Result};
use clap::Subcommand;
use time::format_description::well_known::Rfc3339;
use time::OffsetDateTime;

use crate::api::versions::{PatchVersionRequest, RecordSnapshotRequest};
use crate::context::require_session;
use crate::cryptohelpers::find_file_and_key;
use kutup_crypto::{secretbox, stream};

#[derive(Subcommand)]
pub enum VersionsCmd {
    /// List snapshot versions of a file (newest first).
    List { file_id: String },
    /// Download a specific snapshot version.
    Download {
        file_id: String,
        version_id: String,
        dest: Option<String>,
    },
    /// Restore a snapshot as the latest version (creates a new snapshot).
    Restore { file_id: String, version_id: String },
    /// Set a label on a version (optionally pin it via --keep-forever).
    Label {
        file_id: String,
        version_id: String,
        label: String,
        #[arg(long)]
        keep_forever: bool,
    },
}

pub fn run(profile: &str, json: bool, cmd: &VersionsCmd) -> Result<()> {
    match cmd {
        VersionsCmd::List { file_id } => list(profile, json, file_id),
        VersionsCmd::Download {
            file_id,
            version_id,
            dest,
        } => download(profile, json, file_id, version_id, dest.as_deref()),
        VersionsCmd::Restore {
            file_id,
            version_id,
        } => restore(profile, json, file_id, version_id),
        VersionsCmd::Label {
            file_id,
            version_id,
            label,
            keep_forever,
        } => set_label(profile, json, file_id, version_id, label, *keep_forever),
    }
}

fn list(profile: &str, json: bool, file_id: &str) -> Result<()> {
    let ctx = require_session(profile)?;
    let versions = ctx.client.list_versions(file_id)?;

    if json {
        crate::output::print_json(&versions)?;
        return Ok(());
    }
    if versions.is_empty() {
        println!("(no snapshot versions for this file)");
        return Ok(());
    }
    println!(
        "{}",
        crate::output::header(format!(
            "{:<36}  {:<16}  {:>10}  PIN  LABEL",
            "ID", "CREATED", "SIZE"
        ))
    );
    for v in &versions {
        let pin = if v.keep_forever { "★" } else { " " };
        let label = v.label.as_deref().unwrap_or("");
        println!(
            "{:<36}  {:<16}  {:>10}  {}    {}",
            v.id,
            crate::output::format_time(&v.created_at),
            crate::output::format_bytes(v.size_bytes),
            pin,
            label
        );
    }
    Ok(())
}

fn download(
    profile: &str,
    json: bool,
    file_id: &str,
    version_id: &str,
    dest: Option<&str>,
) -> Result<()> {
    let dest_dir = dest.unwrap_or(".");
    let ctx = require_session(profile)?;
    let master_key = ctx.session.master_key_bytes()?;

    let (row, file_key) = find_file_and_key(&ctx.client, &master_key, file_id)?;
    let meta_bytes = secretbox::open_b64(&row.encrypted_metadata, &row.metadata_nonce, &file_key)
        .context("decrypt metadata")?;
    let meta: crate::api::FileMetadata = serde_json::from_slice(&meta_bytes).unwrap_or_default();

    let short = &version_id[..version_id.len().min(8)];
    let dest_path = {
        let p = std::path::Path::new(dest_dir);
        if p.is_dir() {
            p.join(format!("{}.v-{}", meta.name, short))
        } else {
            p.to_path_buf()
        }
    };

    // Stream: bounded memory + a progress bar, like the main `download`.
    let stream = ctx.client.download_version_stream(file_id, version_id)?;
    let bar = crate::output::progress_bar(stream.content_length(), &meta.name);
    let mut out = std::fs::File::create(&dest_path).context("open dest")?;
    let written = match crate::transfer::stream_download(stream, &file_key, &mut out, |n| {
        bar.set_position(n as u64)
    }) {
        Ok(w) => w,
        Err(e) => {
            drop(out);
            let _ = std::fs::remove_file(&dest_path);
            return Err(e).context("decrypt-write");
        }
    };
    bar.finish_and_clear();

    let dest_str = dest_path.to_string_lossy().into_owned();
    if json {
        crate::output::print_json(
            &serde_json::json!({ "fileId": file_id, "versionId": version_id, "size": written, "dest": dest_str }),
        )?;
    } else {
        println!("Downloaded version {short} of {} → {dest_str}", meta.name);
    }
    Ok(())
}

fn restore(profile: &str, json: bool, file_id: &str, version_id: &str) -> Result<()> {
    let ctx = require_session(profile)?;
    let master_key = ctx.session.master_key_bytes()?;
    let (_, file_key) = find_file_and_key(&ctx.client, &master_key, file_id)?;

    // The restored row carries the SOURCE version's collab metadata: those
    // values become the served x-kutup-seq / x-kutup-doc-key-id headers, and
    // the live-collab values the web client would use are unknowable offline.
    // The update log was already truncated to that seq when the source was
    // recorded, so this is truncation-safe.
    let versions = ctx.client.list_versions(file_id)?;
    let src = versions
        .iter()
        .find(|v| v.id == version_id)
        .ok_or_else(|| {
            crate::errors::NotFound(format!("version {version_id} not found for file {file_id}"))
        })?;

    // download chosen version → decrypt → re-encrypt → snapshot-blob → record.
    let encrypted = ctx.client.download_version(file_id, version_id)?;
    let old = stream::decrypt_stream(&encrypted, &file_key).context("decrypt")?;
    let re_encrypted = stream::encrypt_stream(&old, &file_key).context("re-encrypt")?;
    let size = re_encrypted.len() as i64;

    let blob = ctx.client.upload_snapshot_blob(file_id, re_encrypted)?;
    let now = OffsetDateTime::now_utc()
        .format(&Rfc3339)
        .unwrap_or_default();
    let res = ctx.client.record_snapshot(
        file_id,
        &RecordSnapshotRequest {
            s3_version_id: blob.s3_version_id,
            storage_path: blob.storage_path,
            seq_at_snapshot: src.seq_at_snapshot,
            doc_key_id: src.doc_key_id,
            size_bytes: size,
            label: format!("Restored from {now}"),
            keep_forever: false,
        },
    )?;

    let short = &version_id[..version_id.len().min(8)];
    if json {
        crate::output::print_json(
            &serde_json::json!({ "fileId": file_id, "newVersionId": res.id, "restoredFrom": version_id }),
        )?;
    } else {
        println!(
            "Restored: file={file_id} new-version={} (from {short})",
            res.id
        );
    }
    Ok(())
}

fn set_label(
    profile: &str,
    json: bool,
    file_id: &str,
    version_id: &str,
    label: &str,
    keep_forever: bool,
) -> Result<()> {
    let ctx = require_session(profile)?;
    let patch = PatchVersionRequest {
        label: Some(label.to_string()),
        keep_forever: if keep_forever { Some(true) } else { None },
    };
    let row = ctx.client.patch_version(file_id, version_id, &patch)?;

    if json {
        crate::output::print_json(&row)?;
    } else {
        let short = &version_id[..version_id.len().min(8)];
        println!("Labeled version {short} of file {file_id}");
    }
    Ok(())
}
